[SLP] Order clustered load base pointers by ascending offsets

davemgreen opened the pull request:
This attempts to fix a regression from #98025, where the new order of reduction nodes causes later passes to not be able to produce as nice shuffles. The issue boils down to picking an order of [0 1 3 2] for loaded v4i8 values, which meant later parts could not find a simpler ordering for the shuffles given the legal nodes available in AArch64. If instead we make sure they are ordered [0 1 2 3] then everything can fall into place.

In order to produce a better order that is more likely to work in more cases, this patch takes the existing clustered loads and sort the base pointers if there is an order between them. i.e if `V2 == gep (V1, X)` then V1 is sorted before V2.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -4842,10 +4842,27 @@ static bool clusterSortPtrAccesses(ArrayRef<Value *> VL, Type *ElemTy,
+                                    std::pair<Value *, Value *> V2) {
+    const Value *V = V2.second;
+    while(auto *Gep = dyn_cast<GetElementPtrInst>(V)) {
+      if (Gep->getOperand(0) == V1.second)
+        return true;
+      V = Gep->getOperand(0);
+    }
+    return false;
```
Does it provide strict weak ordering?

alexey-bataev reviewed (commented):
Would be good to have a test that shows benefits, current test changes do no demonstrate this

davemgreen reviewed (commented):
Thanks for the review.

> Would be good to have a test that shows benefits, current test changes do no demonstrate this

The phase-ordering test shows the benefit as the final shuffles are much better for AAch64. `[0 2 4 6 ...]` is a zip, where as `[0 3 4 7 ...]` is a difficult shuffle to materialize. And `[0 1 2 3 ...]` is obviously better than `[0 4 1 5 ...]` at the end. I can try to add a comment to the test case.

alexey-bataev reviewed (approved):
LG

davemgreen committed:
[SLP] Order clustered load base pointers by ascending offsets

This attempts to fix a regression from #98025, where the new order of reduction
nodes causes later passes to not be able to produce as nice shuffles. The issue
boils down to picking an order of [0 1 3 2] for loaded v4i8 values, which meant
later parts could not find a simpler ordering for the shuffles given the legal
nodes available in AArch64. If instead we make sure they are ordered [0 1 2 3]
then everything can fall into place.

In order to produce a better order that is more likely to work in more cases,
this patch takes the existing clustered loads and sort the base pointers if
there is an order between them. i.e if `V2 == gep (V1, X)` then V1 is sorted
before V2.

davemgreen commented:
Thanks

davemgreen commented:
/cherry-pick f67fa3be4db6 e0c14c05a3e3 f2d2ae3f5a8e

MaskRay commented:
@davemgreen When building llvm with `-stdlib=libc++ -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG`, `llvm/test/Transforms/SLPVectorizer/AArch64/loadorder.ll` will fail 

```
Stack dump:
0.	Program arguments: opt -S -passes=slp-vectorizer,instcombine -mtriple=aarch64--linux-gnu
1.	Running pass "function(slp-vectorizer,instcombine<max-iterations=1;no-use-loop-info;verify-fixpoint>)" on module "<stdin>"
2.	Running pass "slp-vectorizer" on function "reduce_blockstrided4x4"
```

davemgreen commented:
Ah OK. I'll take a look - thanks.

davemgreen commented:
Hopefully fixed in #101144.
